staging/src/k8s.io/apiserver/pkg/authentication/token/cache/cached_token_authenticator.go

```diff
@@ -105,19 +115,52 @@ func (a *cachedTokenAuthenticator) AuthenticateToken(ctx context.Context, token
 		return record.resp, record.ok, record.err
 	}
 
-	resp, ok, err := a.authenticator.AuthenticateToken(ctx, token)
-	if !a.cacheErrs && err != nil {
-		return resp, ok, err
+	type lookup struct {
+		resp *authenticator.Response
+		ok   bool
 	}
 
-	switch {
-	case ok && a.successTTL > 0:
-		a.cache.set(key, &cacheRecord{resp: resp, ok: ok, err: err}, a.successTTL)
-	case !ok && a.failureTTL > 0:
-		a.cache.set(key, &cacheRecord{resp: resp, ok: ok, err: err}, a.failureTTL)
+	c := a.group.DoChan(key, func() (val interface{}, err error) {
```

Comment:
If a particular key "hangs" indefinitely somehow, all future uses of that key will timeout right?  We pass in the context to `a.authenticator.AuthenticateToken` and I suppose that should just timeout so this should not be an issue.

I think we need some use of `group.Forget(key)`

Reply:
@liggitt fixed the obvious instance of this in https://github.com/kubernetes/kubernetes/pull/83064

Forgetting means that we leak potentially unbound numbers of goroutines, so it's a tradeoff.

Reply:
yeah, I'd depend on the context timeout being effective